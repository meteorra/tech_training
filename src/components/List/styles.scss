.list {
    list-style-type: none;
    margin: 10px;

    li {
        padding: 10px 0;
    }
}
