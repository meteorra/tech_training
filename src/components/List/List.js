import React from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scss';

const List = ({listItems, changeRoute, deleteListItem}) => {

    const _changeRoute = (e, code) => {
        e.preventDefault();
        changeRoute(`/${code}`);
    };

    const component = listItems.length ?
        <ul className="list city-list">
            {
                listItems.map((el) => (
                    <li
                        key={el.id} className="list-item city-list-item">
                        <div className="form-inline">
                            <div className="form-group">
                                <a href="#" onClick={(e) => _changeRoute(e, el.cityCode)}>
                                    {el.cityName}
                                </a>

                                <button
                                className="btn btn-danger pull-right"
                                onClick={() => deleteListItem(el.id)}>
                                x
                            </button>
                            </div>
                        </div>
                    </li>
                ))
            }
        </ul> :
        <p>No cities are available in the region.</p>;


    return (
        <div>
            {component}
        </div>
    );
};

List.propTypes = {
    listItems: PropTypes.array,
    deleteListItem: PropTypes.func,
};

export default List;
