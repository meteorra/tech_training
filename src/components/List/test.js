import React from 'react';
import List from './List';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

const listItemsDefault = [
    { cityName: 'Minsk', cityCode: 'minsk', id: 0, },
    { cityName: 'Brest', cityCode: 'brest', id: 1, }
];

function setup(listItems = listItemsDefault) {
    const props = {
        changeRoute: jest.fn(),
        deleteListItem: jest.fn(),
        listItems,
    };

    const enzymeWrapper = mount(<List {...props} />);
    const render = renderer.create(<List {...props} />);

    return {
        props,
        enzymeWrapper,
        render,
    };
}

describe('List component', () => {

    it('renders a text message "No cities are available in the region." if no listItems are available', () => {

        const { render, } = setup([]);
        expect(render.toJSON()).toMatchSnapshot();
    });

    it('renders correctly a list with 2 items: Minsk, Brest', () => {

        const { render, } = setup();
        expect(render.toJSON()).toMatchSnapshot();
    });

    it('calls changeRoute with "/minsk" cityCode when Minsk listItem is clicked', () => {
        const { props, enzymeWrapper, } = setup();

        enzymeWrapper.find('.list-item a')
            .at(0)
            .simulate('click');

        expect(props.changeRoute).toBeCalledWith('/minsk');
    });

    it('calls deleteListItem with id=1 when Brest list item delete button is clicked', () => {
        const { props, enzymeWrapper, } = setup();

        enzymeWrapper.find('.list-item .btn-danger')
            .at(1)
            .simulate('click');

        expect(props.deleteListItem).toBeCalledWith(1);
    });

});
