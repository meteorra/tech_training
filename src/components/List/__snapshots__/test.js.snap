// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders a text message "No cities are available in the region." if no listItems are available 1`] = `
<div>
  <p>
    No cities are available in the region.
  </p>
</div>
`;

exports[`List component renders correctly a list with 2 items: Minsk, Brest 1`] = `
<div>
  <ul>
    <li
      className="list-item city-list-item"
    >
      <div
        className="form-inline"
      >
        <div
          className="form-group"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            Minsk
          </a>
          <button
            className="btn btn-danger pull-right"
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </div>
    </li>
    <li
      className="list-item city-list-item"
    >
      <div
        className="form-inline"
      >
        <div
          className="form-group"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            Brest
          </a>
          <button
            className="btn btn-danger pull-right"
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
